src/VisualStudio/Core/Def/Implementation/Experimentation/KeybindingResetDetector.cs

```diff
@@ -0,0 +1,378 @@
+            }
+
+            BrowserHelper.StartBrowser(fwLink);
+
+            KeybindingsResetLogger.Log("ExtensionsLink");
+            _workspace.Options = _workspace.Options.WithChangedOption(KeybindingResetOptions.NeedsReset, false);
+        }
+
+        private void NeverShowAgain()
+        {
+            AssertIsForeground();
+
+            _workspace.Options = _workspace.Options.WithChangedOption(KeybindingResetOptions.NeverShowAgain, true)
+                                                   .WithChangedOption(KeybindingResetOptions.NeedsReset, false);
+            KeybindingsResetLogger.Log("NeverShowAgain");
+
+            // The only external references to this object are as callbacks, which are removed by the Shutdown method.
+            Shutdown();
+        }
+
+        private void InfoBarClose()
+        {
+            AssertIsForeground();
+            _infoBarOpen = false;
+        }
+
+        public int QueryStatus(ref Guid pguidCmdGroup, uint cCmds, OLECMD[] prgCmds, IntPtr pCmdText)
+        {
+            // Technically can be called on any thread, though VS will only ever call it on the UI thread.
+            ThisCanBeCalledOnAnyThread();
+            // We don't care about query status, only when the command is actually executed
+            return (int)OLE.Interop.Constants.OLECMDERR_E_NOTSUPPORTED;
+        }
+
+        public int Exec(ref Guid pguidCmdGroup, uint nCmdID, uint nCmdexecopt, IntPtr pvaIn, IntPtr pvaOut)
+        {
+            // Technically can be called on any thread, though VS will only ever call it on the UI thread.
+            ThisCanBeCalledOnAnyThread();
+            if (pguidCmdGroup == ReSharperCommandGroup && nCmdID >= ResumeId && nCmdID <= ToggleSuspendId)
+            {
+                // Don't delay command processing to update resharper status
+                Task.Run(() => InvokeBelowInputPriority(UpdateStateMachine));
+            }
+
+            // No matter the command, we never actually want to respond to it, so always return not supported. We're just monitoring.
+            return (int)OLE.Interop.Constants.OLECMDERR_E_NOTSUPPORTED;
+        }
+
+        private void OnModalStateChanged(object sender, StateChangedEventArgs args)
+        {
+            ThisCanBeCalledOnAnyThread();
+
+            // Only monitor for StateTransitionType.Exit. This will be fired when the shell is leaving a modal state, including
+            // Tools->Options being exited. This will fire more than just on Options close, but there's no harm from running an
+            // extra QueryStatus.
+            if (args.TransitionType == StateTransitionType.Exit)
+            {
+                InvokeBelowInputPriority(UpdateStateMachine);
+            }
+        }
+
+        public void Shutdown()
+        {
+            AssertIsForeground();
+            if (_priorityCommandTargetCookie != VSConstants.VSCOOKIE_NIL)
+            {
+                _oleComponent.Dispose();
```

Comment:
If you threw an exception between registering the priority target and setting up the OleComponent, this could get strange. Consider just having two ifs, one for the priority target handling and one for the OleComponent?

(yes, this is a case where something *already* went bad. But further badness is good to avoid, and if anybody changes the contract at all where OleComponent could be non-null without the cookie being registered...well, then you have a bug.)